### ISSUE TITLE
Investigate Analyze / Model Polling Forever

### ISSUE THREAD

**Issue by rajadain** [WikiWatershed/model-my-watershed#2695]
Users are unable to get Analyze or Model results as the polls keep going forever and timeout.

**Comment by rajadain** [comment]
Cycling the workers fixed the problem. When the new workers were cycled in they went through a large backlog of jobs from the redis queue.

The most likely candidate for the error is the `celeryd` service had failed. Unfortunately, the logs from the Celery workers do not seem to be in Kibana, making tracing the root cause severely difficult.

One immediate recommendation that can be made is to have a health check on the Celery worker which ensures that the `celeryd` service is running, and either restarts it / cycles the worker / notifies developers or ops that it failed. Because we currently do not have health check in place, failures such as this need to be notified by users.

**Comment by rajadain** [comment]
Made #2697 to follow up.

**Comment by rajadain** [comment]
Just saw this again on staging, spent some time with @arottersman debugging it. Here are the problems discovered in this investigation:

1. Silent failures of geoprocessing. There are cases when the geoprocessing service is up (doing `sudo service geoprocessing status` shows it is active), and the `:8090/ping` endpoint works correctly, but sending even a simple request times out.

2. Silent failures of celery. When the above happens, Celery logs the following error:

    ```
    [2018-03-20 09:04:21,540: ERROR/MainProcess] Task handler raised error: TimeLimitExceeded(300.0,)
    Traceback (most recent call last):
      File "/usr/local/lib/python2.7/dist-packages/billiard/pool.py", line 658, in on_hard_timeout
        raise TimeLimitExceeded(job._timeout)
    TimeLimitExceeded: TimeLimitExceeded(300.0,)
    [2018-03-20 09:04:21,540: ERROR/MainProcess] Hard time limit (300.0s) exceeded for apps.modeling.geoprocessing.run[fe3aaa99-2f32-40f1-b6aa-0fc72fe96fe1]
    [2018-03-20 09:04:21,646: ERROR/MainProcess] Process 'ForkPoolWorker-1' pid:1697 exited with 'signal 9 (SIGKILL)'
    ```

   At this time the workers stop receiving any new jobs, but the output of the service status `sudo service celeryd status` says the workers are up.

3. Restarting services doesn't work the same as stopping and starting them. Often times, stopping and starting brings them back when restarting doesn't.

4. `[nlcd_soil]` jobs (TR-55 modeling) do not save their `model_input` to the database. This can make debugging incidences of errors like this:

    ```
    [2018-03-19 16:55:21,540: ERROR/ForkPoolWorker-1] Task eb662b39-5c65-46a1-ab9d-f7d9315bba6c run from job 655061 raised exception: [nlcd_soil] Geoprocessing Error.
    Details: java.lang.ArrayIndexOutOfBoundsException
    Traceback (most recent call last):
      File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 374, in trace_task
        R = retval = fun(*args, **kwargs)
      File "/usr/local/lib/python2.7/dist-packages/celery/app/trace.py", line 629, in __protected_call__
        return self.run(*args, **kwargs)
      File "apps/modeling/tasks.py", line 101, in nlcd_soil
        raise Exception('[nlcd_soil] {}'.format(result['error']))
    Exception: [nlcd_soil] Geoprocessing Error.
    Details: java.lang.ArrayIndexOutOfBoundsException
    ```

   impossible to trace, as they do not save their inputs. This is because the TR-55 modeling code doesn't use the same standardized `start_celery_job` task that everything else uses, since the exact chain depends on a number of caching factors.

5. Differing behavior of geoprocessing service on local vs staging / production. We've never encountered these kinds of errors while working locally, but the timeouts seem to be quite common given the rollbar message patterns.

**Comment by hectcastro** [comment]
Based on these notes, some ideas that came to mind to help narrow down the problem:

- If we believe that Celery is malfunctioning after geoprocessing service timeouts, we can try to make the Celery jobs to talk to an endpoint that hangs for 20-30s (simulating a hanging geoprocessing service); perhaps there are changes we can make so that Celery is more resilient in the face of a geoprocessing service timout
- For the geoprocessing service, restarts requiring an explicit stop and start point to the service being in a corrupt state; in we encounter this again, we can try to use tooling locally to force a thread dump (could provide some details into where things are stuck in the program)
- Logging (INFO and DEBUG levels) would be a good first step toward improving geoprocessing observability
- I'm a little hesitant to push for a health check based solution at the instance level because these instances have a significant warm up penalty (data volumes for RWD); perhaps process action is more appropriate; ideally, we figure out how to make the geoprocessing service exit whenever it malfunctions so that Upstart can restart it automatically

**Comment by arottersman** [comment]
(I am working on #2738 but will try to keep the discussion here with the rest of everyone's comments)

I was able to reproduce the celery failure in my local environment. If you set in your `settings` `CELERY_TASK_TIME_LIMIT = 1` almost any task is sure to exceed it (interestingly I was unable to reproduce this with a server that hangs for >300sec without explicitly setting this; maybe in debug state celery has no default time limit?). Once celery hits the hard time limit, no other tasks will be accepted, ie the worker is not re-spawned.

I have found several references to this sort of problem: https://github.com/celery/celery/issues/2451 has been the most helpful. It somewhat suggests a soft timelimit might help, but I've found that results in the same state. I am going to try setting a timelimit on the request to the geoprocessing service itself, one that is less than the hard limit.
